src/app/state/save-dialog-actions.ts

```diff
@@ -0,0 +1,175 @@
+import { getMVSData } from './actions';
+
+// SaveDialog Actions
+export function openSaveDialog(options: { saveType: SaveType; sessionId?: string }) {
+  const store = getDefaultStore();
+  const story = store.get(StoryAtom);
+
+  const formData: SaveFormData = {
+    title: story.metadata.title || '',
+    description: '',
+    visibility: options.saveType === 'session' ? 'private' : 'public',
+    tags: '',
+  };
+
+  store.set(SaveDialogAtom, {
+    isOpen: true,
+    saveType: options.saveType,
+    sessionId: options.sessionId,
+    isSaving: false,
+    formData,
+  });
+}
+
+export function closeSaveDialog() {
+  const store = getDefaultStore();
+  const current = store.get(SaveDialogAtom);
+  store.set(SaveDialogAtom, { ...current, isOpen: false });
+}
+
+export function updateSaveDialogFormField(field: keyof SaveFormData, value: string) {
+  const store = getDefaultStore();
+  const current = store.get(SaveDialogAtom);
+
+  store.set(SaveDialogAtom, {
+    ...current,
+    formData: { ...current.formData, [field]: value },
+  });
+}
+
+export function setSaveDialogType(saveType: SaveType) {
+  const store = getDefaultStore();
+  const current = store.get(SaveDialogAtom);
+  store.set(SaveDialogAtom, { ...current, saveType });
+}
+
+async function prepareSessionData(story: Story): Promise<Uint8Array> {
+  const container: StoryContainer = {
+    version: 1,
+    story,
+  };
+
+  // Using message pack for efficient encoding
+  const encoded = encodeMsgPack(container);
+  const deflated = await Task.create('Deflate Story Data', async (ctx) => {
+    return await deflate(ctx, encoded, { level: 3 });
+  }).run();
+
+  return deflated;
+}
+
+async function prepareStateData(story: Story): Promise<Uint8Array | string> {
+  const mvsData = await getMVSData(story);
+
+  if (mvsData instanceof Uint8Array) {
+    return mvsData;
+  } else {
+    // Convert to JSON string for API
+    return JSON.stringify(mvsData);
+  }
+}
+
+async function saveToAPI(data: Uint8Array | string, endpoint: string, formData: SaveFormData) {
+  // Determine correct file extension based on endpoint
+  const getFileExtension = (endpoint: string, data: Uint8Array | string) => {
+    if (endpoint === 'session') {
+      return '.msgpack';
+    } else {
+      // For states, use .mvsj for JSON data, .mvsx for binary data
+      return data instanceof Uint8Array ? '.mvsx' : '.mvsj';
+    }
+  };
+
+  const processedTags = formData.tags
+    .split(',')
+    .map((tag) => tag.trim())
+    .filter((tag) => tag.length > 0);
+
+  const requestBody = {
+    title: formData.title.trim(),
+    filename: formData.title.trim() + getFileExtension(endpoint, data),
+    description: formData.description.trim(),
+    visibility: formData.visibility,
+    tags: processedTags,
+    // Handle data based on type
+    data:
+      data instanceof Uint8Array
+        ? btoa(String.fromCharCode(...data)) // Convert Uint8Array to base64 string
```

Comment:
This will not work for large states and it will throw. You need to use to use something like this:

```
const reader = new FileReader();
reader.readAsDataURL(new Blob([bytes], { type }));
const data = await new Promise<string>((resolve, reject) => {
    reader.onload = () => resolve(reader.result as string);
    reader.onerror = () => reject(reader.error);
});
```

and remove the MIME type at start.